ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/KsqlResource.java

```diff
@@ -130,36 +72,30 @@
 public class KsqlResource {
   // CHECKSTYLE_RULES.ON: ClassDataAbstractionCoupling
 
-  private static final Map<Class<? extends Statement>, Handler<Statement>> CUSTOM_EXECUTORS =
-      ImmutableMap.<Class<? extends Statement>, Handler<Statement>>builder()
-          .put(ListTopics.class,
-              castExecutor(KsqlResource::listTopics, ListTopics.class))
-          .put(ListRegisteredTopics.class,
-              castExecutor(KsqlResource::listRegisteredTopics, ListRegisteredTopics.class))
-          .put(ListStreams.class,
-              castExecutor(KsqlResource::listStreams, ListStreams.class))
-          .put(ListTables.class,
-              castExecutor(KsqlResource::listTables, ListTables.class))
-          .put(ListFunctions.class,
-              castExecutor(KsqlResource::listFunctions, ListFunctions.class))
-          .put(ListQueries.class,
-              castExecutor(KsqlResource::listQueries, ListQueries.class))
-          .put(ShowColumns.class,
-              castExecutor(KsqlResource::showColumns, ShowColumns.class))
-          .put(ListProperties.class,
-              castExecutor(KsqlResource::listProperties, ListProperties.class))
-          .put(Explain.class,
-              castExecutor(KsqlResource::explain, Explain.class))
-          .put(DescribeFunction.class,
-              castExecutor(KsqlResource::describeFunction, DescribeFunction.class))
-          .put(SetProperty.class,
-              castExecutor(KsqlResource::executeDdlImmediately, SetProperty.class))
-          .put(UnsetProperty.class,
-              castExecutor(KsqlResource::executeDdlImmediately, UnsetProperty.class))
-          .put(RunScript.class,
-              castExecutor(KsqlResource::distributeStatement, RunScript.class))
-          .put(TerminateQuery.class,
-              castExecutor(KsqlResource::distributeStatement, TerminateQuery.class))
+  private static final List<ParsedStatement> TERMINATE_CLUSTER =
+      new DefaultKsqlParser().parse(TerminateCluster.TERMINATE_CLUSTER_STATEMENT_TEXT);
+
+  private static final Map<Class<? extends Statement>, StatementExecutor> CUSTOM_EXECUTORS =
+      EnumSet.allOf(CustomExecutors.class)
+          .stream()
+          .collect(Collectors.toMap(
+              CustomExecutors::getStatementClass,
+              Function.identity()
+          ));
+
+  private static final Map<Class<? extends Statement>, StatementValidator> CUSTOM_VALIDATORS =
+      EnumSet.allOf(CustomValidators.class)
+          .stream()
+          .collect(Collectors.toMap(
+              CustomValidators::getStatementClass,
+              Function.identity()
+          ));
+
+  private static final Set<Class<? extends Statement>> OPTIONAL_VALIDATION =
```

Comment:
- RunScript is now handled implicitly by a recent PR, (which you've not merged in yet).
- TerminateQuery requires validation - see old `validateTerminateQuery` method.

Probably an easier / cleaner solution would be to work out what set of statement types that have custom executors, but no custom validator, and just add custom validators, even if they are no-ops.  Then you can drop this set and simplify code.

Reply:
Really like this idea - thanks for the suggestion!